File: week-2 d-4/telemetry/logger.py
#!/usr/bin/env python3
# ─────────────────────────────────────────────────────────────────────
# Day 9 / Week 2 Day 4 — Token-Tracking Telemetry CSV Logger (Python)
#
# Logs every API call to a CSV file with full telemetry.
#
# Usage:
#   from logger import Telemetry
#   tel = Telemetry("telemetry.csv")
#   tel.log(provider="openai", model="gpt-4o-mini", prompt_tokens=100, ...)
#   tel.summary()
#
# CSV columns:
#   timestamp, provider, model, prompt_tokens, completion_tokens,
#   total_tokens, cost, latency_ms, status, prompt_type, notes
# ─────────────────────────────────────────────────────────────────────

import csv
import os
from datetime import datetime
from pathlib import Path


class Telemetry:
    """Logs API call telemetry to CSV and provides summary statistics."""

    def __init__(self, file_path="telemetry.csv", verbose=False):
        self.file_path = Path(file_path).resolve()
        self.verbose = verbose
        self.logs = []
        self._ensure_header()

    def _ensure_header(self):
        self.file_path.parent.mkdir(parents=True, exist_ok=True)
        if not self.file_path.exists():
            with open(self.file_path, "w", newline="", encoding="utf-8") as f:
                writer = csv.writer(f)
                writer.writerow([
                    "timestamp", "provider", "model", "prompt_tokens",
                    "completion_tokens", "total_tokens", "cost",
                    "latency_ms", "status", "prompt_type", "notes",
                ])

    def log(
        self,
        provider="unknown",
        model="unknown",
        prompt_tokens=0,
        completion_tokens=0,
        cost=0.0,
        latency_ms=0.0,
        status="success",
        prompt_type="general",
        notes="",
    ):
        """Log a single API call."""
        timestamp = datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%fZ")
        total_tokens = prompt_tokens + completion_tokens

        row = [
            timestamp, provider, model, prompt_tokens,
            completion_tokens, total_tokens, round(cost, 8),
            round(latency_ms, 1), status, prompt_type, notes,
        ]

        # Append to CSV
        with open(self.file_path, "a", newline="", encoding="utf-8") as f:
            writer = csv.writer(f)
            writer.writerow(row)

        # In-memory
        entry = {
            "timestamp": timestamp,
            "provider": provider,
            "model": model,
            "prompt_tokens": prompt_tokens,
            "completion_tokens": completion_tokens,
            "total_tokens": total_tokens,
            "cost": cost,
            "latency_ms": latency_ms,
            "status": status,
            "prompt_type": prompt_type,
            "notes": notes,
        }
        self.logs.append(entry)

        if self.verbose:
            print(
                f"\033[90m[telemetry]\033[0m {provider} · {model} · "
                f"in:{prompt_tokens} out:{completion_tokens} · "
                f"${cost:.6f} · {latency_ms:.0f}ms · {status}"
            )

        return entry

    def summary(self):
        """Print a full summary of all logged calls."""
        if not self.logs:
            print("\033[33mNo telemetry data logged yet.\033[0m")
            return

        total_calls = len(self.logs)
        successful = sum(1 for l in self.logs if l["status"] == "success")
        failed = total_calls - successful
        total_tok = sum(l["total_tokens"] for l in self.logs)
        total_cost = sum(l["cost"] for l in self.logs)
        total_lat = sum(l["latency_ms"] for l in self.logs)

        # Per-provider
        by_provider = {}
        for l in self.logs:
            p = l["provider"]
            if p not in by_provider:
                by_provider[p] = {"calls": 0, "tokens": 0, "cost": 0.0, "latency": 0.0}
            by_provider[p]["calls"] += 1
            by_provider[p]["tokens"] += l["total_tokens"]
            by_provider[p]["cost"] += l["cost"]
            by_provider[p]["latency"] += l["latency_ms"]

        print("")
        print("\033[32m╔════════════════════════════════════════════════╗\033[0m")
        print("\033[32m║     📊 Telemetry Dashboard — Session Summary  ║\033[0m")
        print("\033[32m╚════════════════════════════════════════════════╝\033[0m")
        print(f"  \033[36mTotal calls:\x1b[0m   {total_calls} ({successful} OK, {failed} failed)")
        print(f"  \033[36mTotal tokens:\x1b[0m  {total_tok:,} ({total_tok / 1_000_000:.2f}M)")
        print(f"  \033[36mTotal cost:\x1b[0m    ${total_cost:.6f}")
        print(f"  \033[36mAvg latency:\x1b[0m   {total_lat / total_calls:.0f}ms")
        print(f"  \033[36mLog file:\x1b[0m      {self.file_path}")
        print("")

        print("  \033[33m── Per-Provider Breakdown ──\033[0m\n")
        for prov, data in sorted(by_provider.items()):
            avg_lat = data["latency"] / data["calls"]
            print(
                f"  \033[36m{prov:12}\033[0m  {data['calls']} calls · "
                f"{data['tokens']:>8,} tokens · "
                f"${data['cost']:.4f} · "
                f"{avg_lat:.0f}ms avg"
            )

        print("")
        avg_cost = total_cost / total_calls
        print(f"  \033[33mAverage cost per call:\033[0m ${avg_cost:.6f}")
        print(f"  \033[33mProjected 10K calls:\033[0m  ${avg_cost * 10000:.2f}")
        print("")

    def close(self):
        self.summary()


if __name__ == "__main__":
    import sys
    file_path = sys.argv[1] if len(sys.argv) > 1 else "telemetry.csv"
    tel = Telemetry(file_path)
    tel.summary()

File: week-2 d-4/telemetry/logger.js
#!/usr/bin/env node
// ─────────────────────────────────────────────────────────────────────
// Day 9 / Week 2 Day 4 — Token-Tracking Telemetry CSV Logger
//
// Logs every API call to a CSV file with full telemetry.
// Can be used standalone or imported as a module.
//
// Usage:
//   const telemetry = new Telemetry("telemetry.csv");
//   telemetry.log({ provider, model, promptTokens, completionTokens,
//                   cost, latencyMs, status, promptType, notes });
//   telemetry.summary();    // print summary to console
//   telemetry.close();      // flush & close
//
// CSV columns:
//   timestamp, provider, model, prompt_tokens, completion_tokens,
//   total_tokens, cost, latency_ms, status, prompt_type, notes
// ─────────────────────────────────────────────────────────────────────

"use strict";

const fs = require("fs");
const path = require("path");

const CSV_HEADER =
  "timestamp,provider,model,prompt_tokens,completion_tokens,total_tokens,cost,latency_ms,status,prompt_type,notes";

class Telemetry {
  /**
   * @param {string} filePath  Path to CSV file (default: ./telemetry.csv)
   * @param {boolean} verbose  Print each log entry to console
   */
  constructor(filePath = "./telemetry.csv", verbose = false) {
    this.filePath = path.resolve(filePath);
    this.verbose = verbose;
    this.logs = [];
    this._ensureHeader();
  }

  _ensureHeader() {
    const dir = path.dirname(this.filePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    if (!fs.existsSync(this.filePath)) {
      fs.writeFileSync(this.filePath, CSV_HEADER + "\n", "utf-8");
    }
  }

  /**
   * Log a single API call to CSV and in-memory buffer.
   */
  log({
    provider = "unknown",
    model = "unknown",
    promptTokens = 0,
    completionTokens = 0,
    cost = 0,
    latencyMs = 0,
    status = "success",
    promptType = "general",
    notes = "",
  } = {}) {
    const timestamp = new Date().toISOString();
    const totalTokens = promptTokens + completionTokens;

    // Escape CSV fields (wrap in quotes if contains comma or quote)
    const esc = (v) => {
      const s = String(v);
      if (s.includes(",") || s.includes('"') || s.includes("\n")) {
        return `"${s.replace(/"/g, '""')}"`;
      }
      return s;
    };

    const row = [
      esc(timestamp),
      esc(provider),
      esc(model),
      promptTokens,
      completionTokens,
      totalTokens,
      cost.toFixed(8),
      latencyMs.toFixed(1),
      esc(status),
      esc(promptType),
      esc(notes),
    ].join(",");

    // Append to file
    fs.appendFileSync(this.filePath, row + "\n", "utf-8");

    // Store in memory for summary
    this.logs.push({
      timestamp,
      provider,
      model,
      promptTokens,
      completionTokens,
      totalTokens,
      cost,
      latencyMs,
      status,
      promptType,
      notes,
    });

    if (this.verbose) {
      const ink = `\x1b[36m${promptTokens}\x1b[0m`;
      const outk = `\x1b[33m${completionTokens}\x1b[0m`;
      console.log(
        `\x1b[90m[telemetry]\x1b[0m ${provider} · ${model} · ` +
        `in:${ink} out:${outk} · \$${cost.toFixed(6)} · ${latencyMs.toFixed(0)}ms · ${status}`
      );
    }
  }

  /**
   * Print a summary of all logged calls.
   */
  summary() {
    if (this.logs.length === 0) {
      console.log("\x1b[33mNo telemetry data logged yet.\x1b[0m");
      return;
    }

    const totalCalls = this.logs.length;
    const successful = this.logs.filter((l) => l.status === "success").length;
    const failed = totalCalls - successful;
    const totalTokens = this.logs.reduce((s, l) => s + l.totalTokens, 0);
    const totalCost = this.logs.reduce((s, l) => s + l.cost, 0);
    const totalLatency = this.logs.reduce((s, l) => s + l.latencyMs, 0);

    // Per-provider breakdown
    const byProvider = {};
    for (const l of this.logs) {
      if (!byProvider[l.provider]) {
        byProvider[l.provider] = { calls: 0, tokens: 0, cost: 0, latency: 0 };
      }
      byProvider[l.provider].calls++;
      byProvider[l.provider].tokens += l.totalTokens;
      byProvider[l.provider].cost += l.cost;
      byProvider[l.provider].latency += l.latencyMs;
    }

    console.log("");
    console.log("\x1b[32m╔════════════════════════════════════════════════╗\x1b[0m");
    console.log("\x1b[32m║     📊 Telemetry Dashboard — Session Summary  ║\x1b[0m");
    console.log("\x1b[32m╚════════════════════════════════════════════════╝\x1b[0m");
    console.log(`  \x1b[36mTotal calls:\x1b[0m   ${totalCalls} (${successful} OK, ${failed} failed)`);
    console.log(`  \x1b[36mTotal tokens:\x1b[0m  ${totalTokens.toLocaleString()} (${(totalTokens / 1_000_000).toFixed(2)}M)`);
    console.log(`  \x1b[36mTotal cost:\x1b[0m    \$${totalCost.toFixed(6)}`);
    console.log(`  \x1b[36mAvg latency:\x1b[0m   ${(totalLatency / totalCalls).toFixed(0)}ms`);
    console.log(`  \x1b[36mLog file:\x1b[0m      ${this.filePath}`);
    console.log("");

    console.log(`  \x1b[33m── Per-Provider Breakdown ──\x1b[0m\n`);
    for (const [prov, data] of Object.entries(byProvider)) {
      const avgLat = data.latency / data.calls;
      console.log(
        `  \x1b[36m${prov.padEnd(12)}\x1b[0m  ${data.calls} calls · ` +
        `${data.tokens.toLocaleString().padStart(8)} tokens · ` +
        `\$${data.cost.toFixed(4).padStart(8)} · ` +
        `${avgLat.toFixed(0)}ms avg`
      );
    }
    console.log("");

    // Bottom line
    const avgCostPerCall = totalCost / totalCalls;
    console.log(`  \x1b[33mAverage cost per call:\x1b[0m \$${avgCostPerCall.toFixed(6)}`);
    console.log(`  \x1b[33mProjected 10K calls:\x1b[0m  \$${(avgCostPerCall * 10000).toFixed(2)}`);
    console.log("");
  }

  /**
   * Close the telemetry (write final summary).
   */
  close() {
    this.summary();
  }
}

// ─── CLI Mode ──────────────────────────────────────────────────
// Run standalone: node logger.js <csv-path>
// Prints summary of existing CSV
if (require.main === module) {
  const filePath = process.argv[2] || "./telemetry.csv";
  const tel = new Telemetry(filePath);
  tel.summary();
}

module.exports = { Telemetry };
